src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php

```diff
@@ -313,6 +314,30 @@ public function testAcceptsProvidersAsTraversable()
         $this->assertSame($refreshedUser, $tokenStorage->getToken()->getUser());
     }
 
+    public function testUserHasChangedEventAsBeenSent()
+    {
+        $tokenStorage = new TokenStorage();
+        $refreshedUser = new User('foobar', 'baz');
+
+        $user = new User('foo', 'bar');
+        $session = new Session(new MockArraySessionStorage());
+        $session->set('_security_context_key', serialize(new UsernamePasswordToken($user, '', 'context_key', ['ROLE_USER'])));
+
+        $request = new Request();
+        $request->setSession($session);
+        $request->cookies->set('MOCKSESSID', true);
+
+        $eventDispatcher = new EventDispatcher();
+        $eventDispatcher->addListener(UserHasChangedEvent::class, function ($event) use ($user) {
+            $this->assertEquals($event->getWrappedEvent()->getUser(), $user);
+        });
+
+        $listener = new ContextListener($tokenStorage, [new NotSupportingUserProvider(), new SupportingUserProvider($refreshedUser)], 'context_key', null, $eventDispatcher);
+        $listener(new RequestEvent($this->getMockBuilder(HttpKernelInterface::class)->getMock(), $request, HttpKernelInterface::MASTER_REQUEST));
+
+        $this->assertNotNull($tokenStorage->getToken());
```

Comment:
hmm, broken :)
The event listener does not call `setAuthenticated()` on the token, so the token should be null here as it has remains deauthenticated, which makes me realize that the token passed to the event does not have the refreshed user but the old one as `$token->setUser($refreshedUser)` has not been called, which was ok before this as it was never reused.